### ISSUE TITLE
[BUG] `DOCUMENT(edge._id)` kills subqueries resulting in less elements in final result set

### ISSUE THREAD

**Issue by konsultaner** [arangodb/arangodb#16451]
## My Environment

* __ArangoDB Version__:        3.9.4
* __Deployment Mode__:       Single Server on premis
* __Deployment Strategy__:   Manual Start
* __Configuration__:               default
* __Infrastructure__:               own
* __Operating System__:        EL7
* __Total RAM in your machine__:        128GB
* __Disks in use__:        SSD
* __Used Package__:           RedHat .rpm

## Component, Query & Data

__Affected feature__:
Query - `DOCUMENT(edge._id)` in subqueries
I have fixed 3 queries where it happened. 2 reduced once are posted here.

__AQL query (if applicable)__:
```
FOR client IN Client
SORT client.displayName
LIMIT @offset,@count

RETURN MERGE(client,{
    accountingMethod: FIRST(
        LET accountingMethod = NOT_NULL(FIRST(
            FOR accountingMethod IN HasAccountingMethod
            FILTER accountingMethod._from == client._id
            LIMIT 1
            RETURN DOCUMENT(accountingMethod._to)),
            {}
        )
        RETURN MERGE(accountingMethod,{
            entries:(
                FOR accountingMethodEntry,edge IN OUTBOUND client HasAccountingMethodEntry
                RETURN MERGE(edge,{
                    accountingMethodEntry:MERGE(accountingMethodEntry,{
                    clientEntry: COUNT(
                            FOR nonClientEntry IN HasAccountingMethodEntry
                            FILTER nonClientEntry._from == accountingMethod._id && nonClientEntry._to == accountingMethodEntry._id
                            LIMIT 1
                            RETURN true
                        ) == 0
                    })
                })
            )
        })
    )
})


```
__AQL explain and/or profile (if applicable)__:
__Profile__
```
Query String (1094 chars, cacheable: false):
 FOR client IN Client
 SORT client.displayName
 LIMIT @offset,@count
 RETURN MERGE(client,{
     accountingMethod: FIRST(
         LET accountingMethod = NOT_NULL(FIRST(
             FOR accountingMethod IN HasAccountingMethod
             FILTER accountingMethod._from == client._id
             LIMIT 1
             RETURN DOCUMENT(accountingMethod._to)),
             {}
         )
         RETURN MERGE(accountingMethod,{
             entries:(
                 FOR accountingMethodEntry,edge IN OUTBOUND client HasAccountingMethodEntry
                 RETURN MERGE(edge,{
                     accountingMethodEntry:MERGE(accountingMethodEntry,{
                     clientEntry: COUNT(
                             FOR nonClientEntry IN HasAccountingMethodEntry
                             FILTER nonClientEntry._from == accountingMethod._id && 
 nonClientEntry._to == accountingMethodEntry._id
                             LIMIT 1
                             RETURN true
                       ...

Execution plan:
 Id   NodeType            Calls     Items   Runtime [s]   Comment
  1   SingletonNode           1         1       0.00001   * ROOT
 37   IndexNode               1      1000       0.01239     - FOR client IN Client   /* skiplist index scan */    
  5   LimitNode               1      1000       0.00002       - LIMIT 0, 1000
 40   SubqueryStartNode       2      2000       0.00633       - LET #12 = ( /* subquery begin */
 46   SubqueryStartNode       3      2000       0.00402         - LET #2 = ( /* subquery begin */
 38   IndexNode               2      1996       0.00277           - FOR accountingMethod IN HasAccountingMethod   /* edge index scan, index only, projections: `_to` */    
 11   LimitNode               2      1996       0.00027             - LIMIT 0, 1
 12   CalculationNode         2      1996       0.00838             - LET #18 = DOCUMENT(accountingMethod.`_to`)   /* simple expression */   /* collections used: accountingMethod : HasAccountingMethod */
 35   LimitNode               2      1996       0.00017             - LIMIT 0, 1
 47   SubqueryEndNode         2      1333       0.00073             - RETURN  #18 ) /* subquery end */
 36   LimitNode               2      1333       0.00013         - LIMIT 0, 1
 15   CalculationNode         2      1333       0.00067         - LET accountingMethod = NOT_NULL(FIRST(#2), { })   /* simple expression */
 42   SubqueryStartNode       2      2000       0.00169         - LET #10 = ( /* subquery begin */
 17   TraversalNode           5      4320       0.06476           - FOR accountingMethodEntry  /* vertex */, edge  /* edge */ IN 1..1  /* min..maxPathDepth */ OUTBOUND client /* startnode */  HasAccountingMethodEntry
 44   SubqueryStartNode       8      7307       0.00613             - LET #8 = ( /* subquery begin */
 39   IndexNode            1065   1035379       0.63541               - FOR nonClientEntry IN HasAccountingMethodEntry   /* edge index scan, index only, projections: `_from`, `_to` */    
 20   CalculationNode      1065   1035379       1.16072                 - LET #22 = ((nonClientEntry.`_from` == accountingMethod.`_id`) && (nonClientEntry.`_to` == accountingMethodEntry.`_id`))   /* simple expression */   /* collections used: nonClientEntry : HasAccountingMethodEntry */
 21   FilterNode              8      7153       0.01591                 - FILTER #22
 22   LimitNode               8      7153       0.00063                 - LIMIT 0, 1
 34   CalculationNode         8      7153       0.00108                 - LET #32 = true   /* json expression */   /* const assignment */
 45   SubqueryEndNode         5      4320       0.00217                 - RETURN  #32 ) /* subquery end */
 26   CalculationNode         5      4320       0.02233             - LET #26 = MERGE(edge, { "accountingMethodEntry" : MERGE(accountingMethodEntry, { "clientEntry" : (COUNT(#8) == 0) }) })   /* simple expression */
 43   SubqueryEndNode         2      1333       0.00244             - RETURN  #26 ) /* subquery end */
 29   CalculationNode         2      1333       0.00356         - LET #28 = MERGE(accountingMethod, { "entries" : #10 })   /* simple expression */
 41   SubqueryEndNode         1       666       0.00104         - RETURN  #28 ) /* subquery end */
 32   CalculationNode         1       666       0.00856       - LET #30 = MERGE(client, { "accountingMethod" : FIRST(#12) })   /* simple expression */   /* collections used: client : Client */
 33   ReturnNode              1       666       0.00001       - RETURN #30

Indexes used:
 By   Name            Type       Collection                 Unique   Sparse   Selectivity   Fields              Ranges
 37   idx_145256591   skiplist   Client                     false    false        99.36 %   [ `displayName` ]   *
 38   edge            edge       HasAccountingMethod        false    false        99.99 %   [ `_from` ]         (accountingMethod.`_from` == client.`_id`)
 17   edge            edge       HasAccountingMethodEntry   false    false        33.55 %   [ `_from` ]         base OUTBOUND
 39   edge            edge       HasAccountingMethodEntry   false    false        33.55 %   [ `_from` ]         (nonClientEntry.`_from` == NOT_NULL(FIRST(#2), { }).`_id`)

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections          Options                                  Filter / Prune Conditions
 17  1..1                       HasAccountingMethodEntry  uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   optimize-subqueries
  2   move-calculations-up-2
  3   use-indexes
  4   remove-filter-covered-by-index
  5   use-index-for-sort
  6   remove-unnecessary-calculations-2
  7   move-calculations-down
  8   reduce-extraction-to-projection
  9   splice-subqueries

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Peak Mem [b]   Exec Time [s]
           0            0           0      1035709     467190        8290304         1.96406

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00002
 parsing                    0.00016
 optimizing ast             0.00002
 loading collections        0.00001
 instantiating plan         0.00014
 optimizing plan            0.00128
 executing                  1.96242
 finalizing                 0.00009
```

__Explain__

```
Query String (1094 chars, cacheable: false):
 FOR client IN Client
 SORT client.displayName
 LIMIT @offset,@count
 RETURN MERGE(client,{
     accountingMethod: FIRST(
         LET accountingMethod = NOT_NULL(FIRST(
             FOR accountingMethod IN HasAccountingMethod
             FILTER accountingMethod._from == client._id
             LIMIT 1
             RETURN DOCUMENT(accountingMethod._to)),
             {}
         )
         RETURN MERGE(accountingMethod,{
             entries:(
                 FOR accountingMethodEntry,edge IN OUTBOUND client HasAccountingMethodEntry
                 RETURN MERGE(edge,{
                     accountingMethodEntry:MERGE(accountingMethodEntry,{
                     clientEntry: COUNT(
                             FOR nonClientEntry IN HasAccountingMethodEntry
                             FILTER nonClientEntry._from == accountingMethod._id && 
 nonClientEntry._to == accountingMethodEntry._id
                             LIMIT 1
                             RETURN true
                       ...

Execution plan:
 Id   NodeType             Est.   Comment
  1   SingletonNode           1   * ROOT
 37   IndexNode           86182     - FOR client IN Client   /* skiplist index scan */    
  5   LimitNode            1000       - LIMIT 0, 1000
 40   SubqueryStartNode    1000       - LET #12 = ( /* subquery begin */
 46   SubqueryStartNode    1000         - LET #2 = ( /* subquery begin */
 38   IndexNode            1000           - FOR accountingMethod IN HasAccountingMethod   /* edge index scan, index only, projections: `_to` */    
 11   LimitNode               1             - LIMIT 0, 1
 12   CalculationNode         1             - LET #18 = DOCUMENT(accountingMethod.`_to`)   /* simple expression */   /* collections used: accountingMethod : HasAccountingMethod */
 35   LimitNode               1             - LIMIT 0, 1
 47   SubqueryEndNode      1000             - RETURN  #18 ) /* subquery end */
 36   LimitNode               1         - LIMIT 0, 1
 15   CalculationNode         1         - LET accountingMethod = NOT_NULL(FIRST(#2), { })   /* simple expression */
 42   SubqueryStartNode       1         - LET #10 = ( /* subquery begin */
 17   TraversalNode           2           - FOR accountingMethodEntry  /* vertex */, edge  /* edge */ IN 1..1  /* min..maxPathDepth */ OUTBOUND client /* startnode */  HasAccountingMethodEntry
 44   SubqueryStartNode       2             - LET #8 = ( /* subquery begin */
 39   IndexNode               4               - FOR nonClientEntry IN HasAccountingMethodEntry   /* edge index scan, index only, projections: `_from`, `_to` */    
 20   CalculationNode         4                 - LET #22 = ((nonClientEntry.`_from` == accountingMethod.`_id`) && (nonClientEntry.`_to` == accountingMethodEntry.`_id`))   /* simple expression */   /* collections used: nonClientEntry : HasAccountingMethodEntry */
 21   FilterNode              4                 - FILTER #22
 22   LimitNode               1                 - LIMIT 0, 1
 34   CalculationNode         1                 - LET #32 = true   /* json expression */   /* const assignment */
 45   SubqueryEndNode         2                 - RETURN  #32 ) /* subquery end */
 26   CalculationNode         2             - LET #26 = MERGE(edge, { "accountingMethodEntry" : MERGE(accountingMethodEntry, { "clientEntry" : (COUNT(#8) == 0) }) })   /* simple expression */
 43   SubqueryEndNode         1             - RETURN  #26 ) /* subquery end */
 29   CalculationNode         1         - LET #28 = MERGE(accountingMethod, { "entries" : #10 })   /* simple expression */
 41   SubqueryEndNode      1000         - RETURN  #28 ) /* subquery end */
 32   CalculationNode      1000       - LET #30 = MERGE(client, { "accountingMethod" : FIRST(#12) })   /* simple expression */   /* collections used: client : Client */
 33   ReturnNode           1000       - RETURN #30

Indexes used:
 By   Name            Type       Collection                 Unique   Sparse   Selectivity   Fields              Ranges
 37   idx_145256591   skiplist   Client                     false    false        99.36 %   [ `displayName` ]   *
 38   edge            edge       HasAccountingMethod        false    false        99.99 %   [ `_from` ]         (accountingMethod.`_from` == client.`_id`)
 17   edge            edge       HasAccountingMethodEntry   false    false        33.55 %   [ `_from` ]         base OUTBOUND
 39   edge            edge       HasAccountingMethodEntry   false    false        33.55 %   [ `_from` ]         (nonClientEntry.`_from` == NOT_NULL(FIRST(#2), { }).`_id`)

Functions used:
 Name       Deterministic   Cacheable   Uses V8
 DOCUMENT   false           false       false  
 FIRST      true            true        false  
 NOT_NULL   true            true        false  
 COUNT      true            true        false  
 MERGE      true            true        false  

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections          Options                                  Filter / Prune Conditions
 17  1..1                       HasAccountingMethodEntry  uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   optimize-subqueries
  2   move-calculations-up-2
  3   use-indexes
  4   remove-filter-covered-by-index
  5   use-index-for-sort
  6   remove-unnecessary-calculations-2
  7   move-calculations-down
  8   reduce-extraction-to-projection
  9   splice-subqueries

Optimization rules with highest execution times:
 RuleName                                    Duration [s]
 use-indexes                                      0.00010
 reduce-extraction-to-projection                  0.00006
 splice-subqueries                                0.00006
 optimize-subqueries                              0.00005
 remove-unnecessary-calculations-2                0.00004

41 rule(s) executed, 1 plan(s) created
```
__Profile on Arango 3.4.x__

```
Query String (1092 chars, cacheable: false):
 FOR client IN Client
 SORT client.displayName
 LIMIT @offset,@count
 RETURN MERGE(client,{
     accountingMethod: FIRST(
         LET accountingMethod = NOT_NULL(FIRST(
             FOR accountingMethod IN HasAccountingMethod
             FILTER accountingMethod._from == client._id
             LIMIT 1
             RETURN DOCUMENT(accountingMethod._to)),
             {}
         )
         RETURN MERGE(accountingMethod,{
             entries:(
                 FOR accountingMethodEntry,edge IN OUTBOUND client HasAccountingMethodEntry
                 RETURN MERGE(edge,{
                     accountingMethodEntry:MERGE(accountingMethodEntry,{
                     clientEntry: COUNT(
                             FOR nonClientEntry IN HasAccountingMethodEntry
                             FILTER nonClientEntry._from == accountingMethod._id && 
 nonClientEntry._to == accountingMethodEntry._id
                             LIMIT 1
                             RETURN true
                       ...

Execution plan:
 Id   NodeType          Calls     Items   Runtime [s]   Comment
  1   SingletonNode         1         1       0.00000   * ROOT
 37   IndexNode             1      1000       0.00857     - FOR client IN Client   /* skiplist index scan */
  5   LimitNode             1      1000       0.00001       - LIMIT 0, 1000
 31   SubqueryNode          1      1000       3.09236       - LET #12 = ...   /* subquery */
  6   SingletonNode      1000      1000       0.00078         * ROOT
 14   SubqueryNode       1000      1000       0.03257           - LET #2 = ...   /* subquery */
  7   SingletonNode      1000      1000       0.00067             * ROOT
 38   IndexNode          1000       995       0.00914               - FOR accountingMethod IN HasAccountingMethod   /* edge index scan, index only, projections: `_to` */
 11   LimitNode          1000       995       0.00102                 - LIMIT 0, 1
 12   CalculationNode    1000       995       0.01656                 - LET #18 = DOCUMENT(accountingMethod.`_to`)   /* simple expression */   /* collections used: accountingMethod : HasAccountingMethod */
 34   LimitNode          1000       995       0.00099                 - LIMIT 0, 1
 13   ReturnNode         1000       995       0.00120                 - RETURN #18
 15   CalculationNode    1000      1000       0.00157           - LET accountingMethod = NOT_NULL(FIRST(#2), { })   /* simple expression */
 28   SubqueryNode       1000      1000       3.04505           - LET #10 = ...   /* subquery */
 16   SingletonNode      1000      1000       0.00117             * ROOT
 17   TraversalNode      1000      4483       0.08711               - FOR accountingMethodEntry  /* vertex */, edge  /* edge */ IN 1..1  /* min..maxPathDepth */ OUTBOUND client /* startnode */  HasAccountingMethodEntry
 25   SubqueryNode       1000      4483       2.90791                 - LET #8 = ...   /* subquery */
 18   SingletonNode      4483      4483       0.00493                   * ROOT
 39   IndexNode          4483   1647694       1.08602                     - FOR nonClientEntry IN HasAccountingMethodEntry   /* edge index scan, index only, projections: `_to`, `_from` */
 20   CalculationNode    4483   1647694       1.77391                       - LET #22 = ((nonClientEntry.`_from` == accountingMethod.`_id`) && (nonClientEntry.`_to` == accountingMethodEntry.`_id`))   /* simple expression */   /* collections used: nonClientEntry : HasAccountingMethodEntry */
 21   FilterNode         4483      4191       0.01633                       - FILTER #22
 22   LimitNode          4483      4191       0.00406                       - LIMIT 0, 1
 35   CalculationNode    4483      4191       0.00369                       - LET #32 = true   /* json expression */   /* const assignment */
 24   ReturnNode         4483      4191       0.00455                       - RETURN #32
 26   CalculationNode    1000      4483       0.04212                 - LET #26 = MERGE(edge, { "accountingMethodEntry" : MERGE(accountingMethodEntry, { "clientEntry" : (COUNT(#8) == 0) }) })   /* simple expression */
 27   ReturnNode         1000      4483       0.00213                 - RETURN #26
 36   LimitNode          1000      1000       0.00090           - LIMIT 0, 1
 29   CalculationNode    1000      1000       0.00729           - LET #28 = MERGE(accountingMethod, { "entries" : #10 })   /* simple expression */
 30   ReturnNode         1000      1000       0.00131           - RETURN #28
 32   CalculationNode       1      1000       0.01313       - LET #30 = MERGE(client, { "accountingMethod" : FIRST(#12) })   /* simple expression */   /* collections used: client : Client */
 33   ReturnNode            1      1000       0.00001       - RETURN #30

Indexes used:
 By   Type       Collection                 Unique   Sparse   Selectivity   Fields              Ranges
 37   skiplist   Client                     false    false        98.41 %   [ `displayName` ]   *
 38   edge       HasAccountingMethod        false    false        99.96 %   [ `_from` ]         (accountingMethod.`_from` == "Client/31041")
 17   edge       HasAccountingMethodEntry   false    false            n/a   [ `_from` ]         base OUTBOUND
 39   edge       HasAccountingMethodEntry   false    false        55.16 %   [ `_from` ]         (nonClientEntry.`_from` == "AccountingMethod/default100z1")

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections          Options                                  Filter / Prune Conditions
 17  1..1                       HasAccountingMethodEntry  uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   optimize-subqueries
  2   move-calculations-up-2
  3   use-indexes
  4   remove-filter-covered-by-index
  5   use-index-for-sort
  6   remove-unnecessary-calculations-2
  7   move-calculations-down
  8   reduce-extraction-to-projection

Query Statistics:
 Writes Exec   Writes Ign   Scan Full   Scan Index   Filtered   Exec Time [s]
           0            0           0      1654172    1643503         3.14755

Query Profile:
 Query Stage           Duration [s]
 initializing               0.00000
 parsing                    0.00012
 optimizing ast             0.00001
 loading collections        0.00001
 instantiating plan         0.00009
 optimizing plan            0.00068
 executing                  3.11432
 finalizing                 0.03231


```

__Explain on Arango 3.4.x__

```
Query String (1092 chars, cacheable: false):
 FOR client IN Client
 SORT client.displayName
 LIMIT @offset,@count
 RETURN MERGE(client,{
     accountingMethod: FIRST(
         LET accountingMethod = NOT_NULL(FIRST(
             FOR accountingMethod IN HasAccountingMethod
             FILTER accountingMethod._from == client._id
             LIMIT 1
             RETURN DOCUMENT(accountingMethod._to)),
             {}
         )
         RETURN MERGE(accountingMethod,{
             entries:(
                 FOR accountingMethodEntry,edge IN OUTBOUND client HasAccountingMethodEntry
                 RETURN MERGE(edge,{
                     accountingMethodEntry:MERGE(accountingMethodEntry,{
                     clientEntry: COUNT(
                             FOR nonClientEntry IN HasAccountingMethodEntry
                             FILTER nonClientEntry._from == accountingMethod._id && 
 nonClientEntry._to == accountingMethodEntry._id
                             LIMIT 1
                             RETURN true
                       ...

Execution plan:
 Id   NodeType           Est.   Comment
  1   SingletonNode         1   * ROOT
 37   IndexNode         86200     - FOR client IN Client   /* skiplist index scan */
  5   LimitNode          1000       - LIMIT 0, 1000
 31   SubqueryNode       1000       - LET #12 = ...   /* subquery */
  6   SingletonNode         1         * ROOT
 14   SubqueryNode          1           - LET #2 = ...   /* subquery */
  7   SingletonNode         1             * ROOT
 38   IndexNode             1               - FOR accountingMethod IN HasAccountingMethod   /* edge index scan, index only, projections: `_to` */
 11   LimitNode             1                 - LIMIT 0, 1
 12   CalculationNode       1                 - LET #18 = DOCUMENT(accountingMethod.`_to`)   /* simple expression */   /* collections used: accountingMethod : HasAccountingMethod */
 34   LimitNode             1                 - LIMIT 0, 1
 13   ReturnNode            1                 - RETURN #18
 15   CalculationNode       1           - LET accountingMethod = NOT_NULL(FIRST(#2), { })   /* simple expression */
 28   SubqueryNode          1           - LET #10 = ...   /* subquery */
 16   SingletonNode         1             * ROOT
 17   TraversalNode         1               - FOR accountingMethodEntry  /* vertex */, edge  /* edge */ IN 1..1  /* min..maxPathDepth */ OUTBOUND client /* startnode */  HasAccountingMethodEntry
 25   SubqueryNode          1                 - LET #8 = ...   /* subquery */
 18   SingletonNode         1                   * ROOT
 39   IndexNode             1                     - FOR nonClientEntry IN HasAccountingMethodEntry   /* edge index scan, index only, projections: `_to`, `_from` */
 20   CalculationNode       1                       - LET #22 = ((nonClientEntry.`_from` == accountingMethod.`_id`) && (nonClientEntry.`_to` == accountingMethodEntry.`_id`))   /* simple expression */   /* collections used: nonClientEntry : HasAccountingMethodEntry */
 21   FilterNode            1                       - FILTER #22
 22   LimitNode             1                       - LIMIT 0, 1
 35   CalculationNode       1                       - LET #32 = true   /* json expression */   /* const assignment */
 24   ReturnNode            1                       - RETURN #32
 26   CalculationNode       1                 - LET #26 = MERGE(edge, { "accountingMethodEntry" : MERGE(accountingMethodEntry, { "clientEntry" : (COUNT(#8) == 0) }) })   /* simple expression */
 27   ReturnNode            1                 - RETURN #26
 36   LimitNode             1           - LIMIT 0, 1
 29   CalculationNode       1           - LET #28 = MERGE(accountingMethod, { "entries" : #10 })   /* simple expression */
 30   ReturnNode            1           - RETURN #28
 32   CalculationNode    1000       - LET #30 = MERGE(client, { "accountingMethod" : FIRST(#12) })   /* simple expression */   /* collections used: client : Client */
 33   ReturnNode         1000       - RETURN #30

Indexes used:
 By   Type       Collection                 Unique   Sparse   Selectivity   Fields              Ranges
 37   skiplist   Client                     false    false        98.41 %   [ `displayName` ]   *
 38   edge       HasAccountingMethod        false    false        99.96 %   [ `_from` ]         (accountingMethod.`_from` == client.`_id`)
 17   edge       HasAccountingMethodEntry   false    false            n/a   [ `_from` ]         base OUTBOUND
 39   edge       HasAccountingMethodEntry   false    false        55.16 %   [ `_from` ]         (nonClientEntry.`_from` == NOT_NULL(FIRST(#2), { }).`_id`)

Functions used:
 Name       Deterministic   Cacheable   Uses V8
 DOCUMENT   false           false       false  
 FIRST      true            true        false  
 NOT_NULL   true            true        false  
 COUNT      true            true        false  
 MERGE      true            true        false  

Traversals on graphs:
 Id  Depth  Vertex collections  Edge collections          Options                                  Filter / Prune Conditions
 17  1..1                       HasAccountingMethodEntry  uniqueVertices: none, uniqueEdges: path                           

Optimization rules applied:
 Id   RuleName
  1   optimize-subqueries
  2   move-calculations-up-2
  3   use-indexes
  4   remove-filter-covered-by-index
  5   use-index-for-sort
  6   remove-unnecessary-calculations-2
  7   move-calculations-down
  8   reduce-extraction-to-projection


```

__Dataset__:
[debugDump.zip](https://github.com/arangodb/arangodb/files/8974777/debugDump.zip)

## Steps to reproduce

__Problem__:

We are currently migrating from Arango 3.4.x to 3.9.x. While Testing everything we found that issue. On the 3.4.x machine the query returned 1000 clients. On the 3.9.2 machine only 666 clients were returned. I think this is a bug. The `MERGE` is done in the `RETURN` and should not affect the result amount. The client collection has enough entries.

__Expected result__:

`LIMIT 0,1000` should return 1000 clients instead of 666

**Comment by konsultaner** [comment]
@dothebart Do you any need more information or data for that issue to be solved?

**Comment by konsultaner** [comment]
@dothebart An equal issue appeared in one of my other queries. A simple `DOCUMENT(anId)` caused to skip the result set entries. replacing it with `{_id: anId}` and materializing it later in the query helped. 

**This is a serious issue!**

It is this part of the query:

```AQL
LET foodIntolerance = (
    FOR ingredient IN NOT_NULL(ingredients,[])
    FOR hasFoodIntoleranceTo IN HasFoodIntoleranceTo
    FILTER hasFoodIntoleranceTo._from == client._id
    COLLECT allergenId = hasFoodIntoleranceTo._to INTO allergenInfo
    //RETURN DOCUMENT(allergenId) // result 2858 <-this probably causes a sub query to interrupt ?
    RETURN {_id:allergenId} // result 3117
)
```

This is the other query as a debug package:
[debugPackage2.zip](https://github.com/arangodb/arangodb/files/9961272/debugPackage2.zip)

**Comment by konsultaner** [comment]
@dothebart Both queries try to call `DOCUMENT(edge._id)` maybe they run into an error where the subquery is canceled in the background when trying to gather the edge collection data? I'm currently working on the third query that has the exact same issue. Something is really broken here.

**Comment by konsultaner** [comment]
@hkernbach I cannot send you the whole set since it holds sensible data, but I will prepare a dataset that reproduces the case.

**Comment by Simran-B** [comment]
It is still reproducible but not always and only with the full database as it is or using a complete dump of that database, is that right?

Just as a suggestion, would it be possible to do a screensharing session if you can't share the dump?

**Comment by hkernbach** [comment]
@Simran-B We've gotten a from @konsultaner dataset already and are able to proceed on this.

**Comment by hkernbach** [comment]
@konsultaner Thank you for your help. We could break down and isolate this issues to a bare minimum now. We can confirm, this is a bug - and we're on it to fix it. Your contribution helped a lot :)

**Comment by konsultaner** [comment]
@hkernbach Oh that's good news. I love to help! The best database on the market. You helped me a lot too!

**Comment by Simran-B** [comment]
As there is no backport yet, it will likely be released as part of v3.9.7

**Comment by hkernbach** [comment]
Yes will be part of v3.9.7:

Will link all the related backports here as well: (3.8.X, 3.9.7, 3.10.X)
https://github.com/arangodb/arangodb/pull/17709
https://github.com/arangodb/arangodb/pull/17710
https://github.com/arangodb/arangodb/pull/17711
